### PR TITLE
approle: secret-id listing lock sanity check

### PR DESCRIPTION
**Pull request by vishalnayak**
Ping #2314

### CODE REVIEW

File: builtin/credential/approle/path_role.go [hunk 1]
```diff
@@ -523,9 +523,9 @@ func (b *backend) pathRoleSecretIDList(req *logical.Request, data *framework.Fie
 	// If the argument to secretIDLock does not start with 2 hex
 	// chars, a generic lock is returned. So, passing empty string
 	// to get the "custom" lock that could be used for listing.
-	lock := b.secretIDLock("")
-	lock.RLock()
-	defer lock.RUnlock()
+	customLock := b.secretIDLock("")
+	customLock.RLock()
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Shouldn't this simply not get locked at this point in time? What is it protecting?

File: builtin/credential/approle/path_role.go [hunk 2]
```diff
@@ -550,22 +550,27 @@ func (b *backend) pathRoleSecretIDList(req *logical.Request, data *framework.Fie
 		// corresponding lock many times using SecretIDs is not
 		// possible. Also, indexing it everywhere using secretIDHMACs
 		// makes listing operation easier.
-		lock := b.secretIDLock(secretIDHMAC)
-		lock.RLock()
+		secretIDLock := b.secretIDLock(secretIDHMAC)
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: It seems like the overall effect of this PR is to account for the case where secretIDHMAC == "" in which case maybe there's a better check that can be in place? (Or, the lock doesn't need to be locked earlier?)

File: builtin/credential/approle/path_role.go (hunk 1 above)
Comment (vishalnayak) [inline builtin/credential/approle/path_role.go]: 1) Concurrent list operations.
2) Concurrent list and role deletion operations.

Both are low frequency operations though.

File: builtin/credential/approle/path_role.go (hunk 2 above)
Comment (vishalnayak) [inline builtin/credential/approle/path_role.go]: Yeah. Essentially it is the case where secretIDHMAC == "". But secretIDHMAC is always a valid hex encoded string. So hitting this case is also unlikely. I wasn't able to reason when and how the locks will be same. Also, I can't see a case where this ends up with a deadlock when locks are not the same.

**Comment by vishalnayak** [comment]
Using a dedicated lock for listing secret id accessors, eliminating the possibilities of listing lock being the same as the one used for accessing any of the secret IDs.

**Pull request merged**